[Form] Add support for IntegerType in DateTimeToArrayTransformer

ivanbarlog opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no <!-- don't forget to update src/**/CHANGELOG.md files -->
| BC breaks?    | no
| Deprecations? | no <!-- don't forget to update UPGRADE-*.md files -->
| Tests pass?   | yes
| Fixed tickets | #... <!-- #-prefixed issue number(s), if any -->
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!--highly recommended for new features-->

<!--
- Bug fixes must be submitted against the lowest branch where they apply
  (lowest branches are regularly merged to upper ones so they get the fixes too).
- Features and deprecations must be submitted against the master branch.
- Replace this comment by a description of what your PR is solving.
-->

The `TextType` was working with empty strings as default values but the `IntegerType` was throwing an exception since it was expected the numeric type not a `string`. This way it works with both `TextType` and `IntegerType`.


Not sure if there is particular reason why there were empty strings instead of `null` but hope that this won't break anything.

ivanbarlog committed:
Add support for IntegerType

The `TextType` was working with empty strings as default values but the `IntegerType` was throwing an exception since it was expected the numeric type not a `string`. This way it works with both `TextType` and `IntegerType`.

Simperfit commented:
Travis failure is unrelated.

ivanbarlog commented:
@Simperfit I am not sure why there is still test for 5.4 since the support was already dropped by php.net

What about the pull request? Can you review it?

Simperfit commented:
Because when 2.7 was release it was still supported by php and we cannot drop support of a LTS version like this because people may use  5.4 anyway ;).

ivanbarlog commented:
Right, I haven't realized that. Thanks for clarification :-)

xabbuh commented:
> The `TextType` was working with empty strings as default values but the `IntegerType` was throwing an exception since it was expected the numeric type not a `string`.

Can you add a test that would fail with the mentioned exception? As far as I don't see the modified tests do not cover this.

nicolas-grekas commented:
Status: needs work

ivanbarlog commented:
@xabbuh could you please explain what test I should prepare in more detail? Since I am not sure if I understand ;-)

Thanks

ivanbarlog commented:
So the example is when you are try to use the transformer with the `IntegerType` as you can see below.

```php
public function buildForm(FormBuilderInterface $builder, array $options)
{
    $builder
        ->add('day', IntegerType::class)
        ->add('month', IntegerType::class)
        ->add('year', IntegerType::class)
        ->addModelTransformer(
            new \Symfony\Component\Form\Extension\Core\DataTransformer\DateTimeToArrayTransformer(
                null, null, ['year', 'month', 'day']
            )
        );
}
```

You will end up with the following error:

> Unable to transform value for property path "[day]": Expected a numeric.

```
Symfony\Component\Form\Exception\TransformationFailedException:
Unable to transform value for property path "[day]": Expected a numeric.

  at vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:1107
  at Symfony\Component\Form\Form->normToView('')
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:350)
  at Symfony\Component\Form\Form->setData('')
     (vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Core/DataMapper/PropertyPathMapper.php:49)
  at Symfony\Component\Form\Extension\Core\DataMapper\PropertyPathMapper->mapDataToForms(array('year' => '', 'month' => '', 'day' => ''), object(RecursiveIteratorIterator))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:383)
  at Symfony\Component\Form\Form->setData(null)
     (vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Core/DataMapper/PropertyPathMapper.php:49)
  at Symfony\Component\Form\Extension\Core\DataMapper\PropertyPathMapper->mapDataToForms(object(QuestionContainer), object(RecursiveIteratorIterator))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:383)
  at Symfony\Component\Form\Form->setData(object(QuestionContainer))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:487)
  at Symfony\Component\Form\Form->initialize()
     (vendor/symfony/symfony/src/Symfony/Component/Form/FormBuilder.php:226)
  at Symfony\Component\Form\FormBuilder->getForm()
     (vendor/symfony/symfony/src/Symfony/Component/Form/FormFactory.php:30)
  at Symfony\Component\Form\FormFactory->create('AppBundle\\Form\\Type\\GenericType', object(QuestionContainer), array())
     (src/BusinessLogic/Questionnaire/FormHandler/QuestionnaireStepBuilder.php:55)
  at BusinessLogic\Questionnaire\FormHandler\QuestionnaireStepBuilder->createForm(object(DateOfBirthConfig), object(InterestedParty), array())
     (src/BusinessLogic/Questionnaire/FormHandler/FormFactory.php:43)
  at BusinessLogic\Questionnaire\FormHandler\FormFactory->createForm(object(DateOfBirthConfig), object(InterestedParty))
     (src/BusinessLogic/Questionnaire/FormHandler/EditInterestedPartyFormHandler.php:55)
  at BusinessLogic\Questionnaire\FormHandler\EditInterestedPartyFormHandler->handleRequest(object(Step))
     (src/AppBundle/Controller/Client/LicenceApplication/InterestedParty/EditController.php:79)
  at AppBundle\Controller\Client\LicenceApplication\InterestedParty\EditController->executeAction(object(Request), 'LBN-100001-1', 4, 'date-of-birth')
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:151)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:68)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (web/app_dev.php:12)
```

```
Symfony\Component\Form\Exception\TransformationFailedException:
Expected a numeric.

  at vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Core/DataTransformer/NumberToLocalizedStringTransformer.php:113
  at Symfony\Component\Form\Extension\Core\DataTransformer\NumberToLocalizedStringTransformer->transform('')
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:1104)
  at Symfony\Component\Form\Form->normToView('')
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:350)
  at Symfony\Component\Form\Form->setData('')
     (vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Core/DataMapper/PropertyPathMapper.php:49)
  at Symfony\Component\Form\Extension\Core\DataMapper\PropertyPathMapper->mapDataToForms(array('year' => '', 'month' => '', 'day' => ''), object(RecursiveIteratorIterator))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:383)
  at Symfony\Component\Form\Form->setData(null)
     (vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Core/DataMapper/PropertyPathMapper.php:49)
  at Symfony\Component\Form\Extension\Core\DataMapper\PropertyPathMapper->mapDataToForms(object(QuestionContainer), object(RecursiveIteratorIterator))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:383)
  at Symfony\Component\Form\Form->setData(object(QuestionContainer))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:487)
  at Symfony\Component\Form\Form->initialize()
     (vendor/symfony/symfony/src/Symfony/Component/Form/FormBuilder.php:226)
  at Symfony\Component\Form\FormBuilder->getForm()
     (vendor/symfony/symfony/src/Symfony/Component/Form/FormFactory.php:30)
  at Symfony\Component\Form\FormFactory->create('AppBundle\\Form\\Type\\GenericType', object(QuestionContainer), array())
     (src/BusinessLogic/Questionnaire/FormHandler/QuestionnaireStepBuilder.php:55)
  at BusinessLogic\Questionnaire\FormHandler\QuestionnaireStepBuilder->createForm(object(DateOfBirthConfig), object(InterestedParty), array())
     (src/BusinessLogic/Questionnaire/FormHandler/FormFactory.php:43)
  at BusinessLogic\Questionnaire\FormHandler\FormFactory->createForm(object(DateOfBirthConfig), object(InterestedParty))
     (src/BusinessLogic/Questionnaire/FormHandler/EditInterestedPartyFormHandler.php:55)
  at BusinessLogic\Questionnaire\FormHandler\EditInterestedPartyFormHandler->handleRequest(object(Step))
     (src/AppBundle/Controller/Client/LicenceApplication/InterestedParty/EditController.php:79)
  at AppBundle\Controller\Client\LicenceApplication\InterestedParty\EditController->executeAction(object(Request), 'LBN-100001-1', 4, 'date-of-birth')
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:151)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:68)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (web/app_dev.php:12)
```

I think it's caused by the `IntegerType` which can accept only numeric data or null so setting `null` instead of empty string within the transformer is working. 

Do you think that changing these values from empty string to `null` may affect any other case?

xabbuh commented:
Now that I understand your issue I would at least say that this is not a bugfix but if we want to support it, this would be a new feature.

Then, we must not change existing tests as it would mean that we won't ever notice any changes breaking the formerly supported usage of the transformer.

ivanbarlog commented:
Sorry that I didn't reported the whole error earlier. So what should we do next? Should I somehow change my pull request? Or it is just the change from bugfix to feature?

xabbuh commented:
I am actually not sure if that's really something we need to support in core or if you wouldn't be better off by decorating the existing transformer thus being able to customise it for your needs.

ivanbarlog commented:
Well, right now the transformer works only with string form types eg. `TextType`. I've added support for `IntegerType` and it still works perfectly with `TextType` so I guess it's OK to add it to the repository. Right now I have copied the Transformer within my project and replaced few empty strings with `null`. But I think that's really shitty solution and I don't see any other better solution for this... I don't see how you would like to decorate this since it's internal workings of the transformer. If you have an example I would be flattered to use it within my project...

xabbuh commented:
@ivanbarlog Please apologise for not coming back to you with this earlier. I have thought about this again and I think we cannot make this change. For everyone relying on empty strings as the result here (which is what you would expect as empty view data for a text type) this will be a breaking change.

In your own project I suggest not to copy the built-in date transformer, but to add transformers to the nested integer types to convert the empty string to `null` there. This way, you don't have to maintain a copy of the built-in transformer and can still use the form type like you would like to do.

That being said I am going to close here. Once again sorry for the delay and that your contribution didn't make it this time. I hope to see you again in the future with some other ideas.
